packages/grafana-llm-app/src/components/AppConfig/DevSandbox/DevSandboxToolInspector.tsx

```diff
@@ -269,6 +386,30 @@ export function DevSandboxToolInspector({ tools }: DevSandboxToolInspectorProps)
     return name.includes(searchLower) || description.includes(searchLower);
   });
 
+  // Group tools by category
+  const groupedTools = filteredTools.reduce(
+    (groups, tool) => {
+      const category = TOOL_CATEGORIES[tool.name] || 'Ungrouped';
+      if (!groups[category]) {
+        groups[category] = [];
+      }
+      groups[category].push(tool);
+      return groups;
+    },
+    {} as Record<string, Tool[]>
+  );
+
+  // Sort categories with Ungrouped at the end
+  const sortedCategories = Object.keys(groupedTools).sort((a, b) => {
+    if (a === 'Ungrouped') {
+      return 1;
+    }
+    if (b === 'Ungrouped') {
+      return -1;
+    }
+    return a.localeCompare(b);
+  });
```

Comment:
I'm not exactly sure what it is but something is causing this modal to run really slowly for me, perhaps this needs memoizing? not sure... a profiler didn't actually pick anything up so maybe it was temporary.

Reply:
hmm [this profile](https://share.firefox.dev/474OcFK) demonstrates lots of high cpu usage, seems related to `<Input>`s

Reply:
Yeah I think it's something with the `ToolParameterForm` causing lots of rerenders in an infinite loop, maybe one of the useEffects.